docs/proposals/scheduling/20230415-numa-topology-scheduling.md

```diff
@@ -0,0 +1,238 @@
+--- a/apis/scheduling/v1alpha1/device_types.go
++++ b/apis/scheduling/v1alpha1/device_types.go
+@@ -44,6 +44,15 @@ type DeviceInfo struct {
+type DeviceInfo struct {
+	// UUID represents the UUID of device
+	UUID string `json:"id,omitempty"`
+	// Minor represents the Minor number of Device, starting from 0
+	Minor *int32 `json:"minor,omitempty"`
+	// Type represents the type of device
+	Type DeviceType `json:"type,omitempty"`
+	// Health indicates whether the device is normal
+	Health bool `json:"health,omitempty"`
+	// Resources is a set of (resource name, quantity) pairs
+	Resources corev1.ResourceList `json:"resources,omitempty"`
++	// Topology represents the topology information about the device
++	Topology *DeviceTopology `json:"topology,omitempty"`
++}
++
++type DeviceTopology struct {
++	SocketID int32  `json:"socketID"`
++	NodeID   int32  `json:"nodeID"`
++	PCIEID   int32  `json:"pcieID"`
++	BusID    string `json:"busID,omitempty"`
+ }
+```
+
+### Changes in koordlet
+
+Because we need to divide the CPU shared pool into multiple instances according to the NUMA Topology, some of the elasticity capabilities of koordlet need to be adapted. For example, when dynamically adjusting more CPUs for the CPUShare Pod, it should only be processed in the specified NUMA Node, rather than on the entire machine. Of course, this is not strict for a single-machine context. We can consider whether the Pod has device resources. If it does, it is best not to cross NUMA Nodes, otherwise it may cause some performance issues.
+
+If the current NUMA Topology Policy of the node is BestEffort, we can also consider handling it according to the original elastic mechanism, and only enhance the restrictions when it is Restricted/SingleNUMANode.
+
+In the reporting logic of the Device CRD, we need to adapt to support new fields. When reporting the NodeResourceTopology, we need to report the NUMA Topology policy of kubelet and write it to the field NodeResourceTopology.TopologyPolicies according to the NodeResourceTopology specification.
+
+### Compatibility
+
+- Because Koordinator Scheduler adopts the same Preferred NUMA Node Affinity calculation method as kubelet Topology Manager, and Koordinator Scheduler can perceive the Topology Policy configured by kubelet, theoretically, the two can maintain an identical logic, so there will no longer be TopologyError-related errors even when Topology Manager is enabled on the node. However, in a production environment, it is still advisable to consider disabling kubelet's Topology Manager to avoid unknown risks as much as possible.
+- Koordinator Scheduler only supports resource allocation at the Pod level, so if kubelet is configured at the Container level, conflicts may occur.
+- Koordinator's NUMA Topology Scheduling can only be applied to newly imported nodes. When there are existing nodes and Pods running on them, it is best to clear the Pods on the node before configuring the NUMA Topology Policy.
```

Comment:
Is it a strict requirement that we must drain out all pods on node before we apply koord-topo-policy on Node?

Otherwise, if there are already some pods allocated by kubelet, koordlet should report the allocated info on NodeTopo CR to let koord-scheduler knows it.

Reply:
Yes, before applying the NUMA policy, the user needs to ensure that there are no other Pods on the node (except DaemonSet).